tests/test_wheel.py

```diff
@@ -191,12 +191,15 @@ def test_rpath(wheel_link_against_local_lib, tmp_path):
     artifact = wheel.wheelfile.WheelFile(wheel_link_against_local_lib)
     artifact.extractall(tmp_path)
 
+    module = tmp_path / f'example{EXT_SUFFIX}'
     if platform.system() == 'Linux':
-        elf = mesonpy._elf.ELF(tmp_path / f'example{EXT_SUFFIX}')
-        assert '$ORIGIN/.link_against_local_lib.mesonpy.libs' in elf.rpath
-    else:  # 'Darwin'
-        dylib = mesonpy._dylib.Dylib(tmp_path / f'example{EXT_SUFFIX}')
-        assert '@loader_path/.link_against_local_lib.mesonpy.libs' in dylib.rpath
+        binary = mesonpy._elf.ELF(module)
+        libdir_origin = '$ORIGIN'
+    elif platform.system() == 'Darwin':
+        binary = mesonpy._dylib.Dylib(module)
+        libdir_origin = '@loader_path'
+
+    assert binary.rpath == {f'{libdir_origin}/.link_against_local_lib.mesonpy.libs'}
```

Comment:
Testing on macOS shows a difference, there the test does pass within a conda env and there's only one RPATH entry with this PR:
```
(Pdb) binary.rpath
frozenset({'@loader_path/.link_against_local_lib.mesonpy.libs'})
```
while on `main` I get:
```
(Pdb) dylib.rpath
frozenset({'/Users/rgommers/mambaforge/envs/scipy-dev/lib', '@loader_path/.link_against_local_lib.mesonpy.libs', '@loader_path/'})
```
The output on `main` is correct here. I did a bit of digging, but it wasn't immediately clear what is happening in `_dylib.py` and the macOS-specific RPATH postprocessing that's not done for Linux.

Reply:
> the path to the conda-forge env's `lib/` should be there

Why? This is a wheel, it's meant to be distributed, it shouldn't be leaking your local environment paths.

> . I did a bit of digging, but it wasn't immediately clear what is happening in `_dylib.py` and the macOS-specific RPATH postprocessing that's not done for Linux.

I assume you meant changing `LC_LOAD_DYLIB`, right? See https://github.com/mesonbuild/meson-python/pull/348#issuecomment-1473618719.

Reply:
To be more concrete, if I have a conda env and I type `pip install .`, I'm going through a wheel and I expect things to work without manually messing with `LD_LIBRARY_PATH`.

Reply:
> This is conceptually wrong, wheels have multiple purposes. See the paths at [pypackaging-native.github.io/meta-topics/build_steps_conceptual/#building-and-installing-or-uploading-artifacts](https://pypackaging-native.github.io/meta-topics/build_steps_conceptual/#building-and-installing-or-uploading-artifacts).

Fair enough.

> That is not correct, the RPATH entry is supposed to be there, and only when you're distributing you must use `auditwheel` to strip it. I'm sure we went over this in other issues.

I don't think we discussed stripping the RPATH specifically.

Is there any particular reason you think the RPATH should be set here?

> To be more concrete, if I have a conda env and I type `pip install .`, I'm going through a wheel and I expect things to work without manually messing with `LD_LIBRARY_PATH`.

Won't lots of other things break in that case? I was under the impression that you needed to enable a conda environment to use it, and setting up the library search path is one of the things enabling an environment is supposed to do, lots of things rely on that. Is this not the case?

Setting the RPATH to a user directory this is problematic for reproducible builds, so I'd like to avoid doing that if possible.

Reply:
No, conda envs are relocatable via RPATH rewriting. Spack envs too I believe. See, e.g., https://github.com/mesonbuild/meson-python/issues/125#issuecomment-1249241305 for details. That shows an example of how conda edits `LDFLAGS` on environment activation to insert `-Wl,-rpath,${CONDA_PREFIX}/lib`. That's why they are there in my binaries, and they should not be stripped by `meson-python`.

> Setting the RPATH to a user directory is problematic for reproducible builds, so I'd like to avoid doing that if possible.

This is not problematic. Reproducible builds are reproducible _for a given environment_, changing things like rpaths is completely normal. The rules here are simple:
1. don't change any RPATH entries in meson-python, except for inserting a `.mesonpy.libs` one in case we have an internal library
2. anyone wanting to distribute wheels must actually do the work to make them distributable by building in the right environment (e.g. a Docker manylinux container) and running a tool like `auditwheel` as appropriate.

You clearly cannot expect to take any random build env and then have `meson-python` spit out a wheel that is okay to upload to PyPI. 

This doesn't only apply to conda by the way, it will apply to any prefix dev env that's outside a standard loader search path (IIRC Nix also does something like this).

Reply:
> This is not problematic. Reproducible builds are reproducible _for a given environment_, changing things like rpaths is completely normal.

It isn't that simple :sweat_smile:

While it is possible to achieve reproducible builds when setting the RPATH, tools to validate reproducibility need to add support for arbitrary build locations to be able to handle these cases. That's why I said it is *problematic*, not that it makes reproducible builds impossible. https://github.com/kushaldas/asaman for eg. does not have support for this.

But you pointed out conda does this via `LDFLAGS`, which makes much more sense to me, and I agree we should be keeping that.

> don't change any RPATH entries in meson-python, except for inserting a `.mesonpy.libs` one in case we have an internal library

I slightly disagree. IMO we should be able to change RPATHs set by Meson directly if it doesn't match the behavior we want, but we should definitely keep anything that comes out of user configurations, like `LDFLAGS`.

> No, conda envs are relocatable via RPATH rewriting. Spack envs too I believe. See, e.g., [#125 (comment)](https://github.com/mesonbuild/meson-python/issues/125#issuecomment-1249241305) for details. That shows an example of how conda edits `LDFLAGS` on environment activation to insert `-Wl,-rpath,${CONDA_PREFIX}/lib`. That's why they are there in my binaries, and they should not be stripped by `meson-python`.

Thank you, this was very helpful to me :+1: 

> You clearly cannot expect to take any random build env and then have `meson-python` spit out a wheel that is okay to upload to PyPI.

I don't think this is an unreasonable expectation, but it definitely depends on the project. We should have more documentation regarding this to help users.

> This doesn't only apply to conda by the way, it will apply to any prefix dev env that's outside a standard loader search path (IIRC Nix also does something like this).

Yes.

IIRC Nix sets `LD_LIBRARY_PATH` on the fly to build very specific environments, but it's possible that they're using RPATH for something.

Reply:
This has now tripped us up several times. I think we should add a separate test for this. Maybe a good way of doing that is using the `install_rpath` argument to [executable](https://mesonbuild.com/Reference-manual_functions.html#executable). It's documented as: _A string to set the target's rpath to after install (but not before that). On Windows, this argument has no effect._. So if we set it to some unique string and check that that string gets preserved, that's probably a fairly robust test.

Reply:
Agreed, I'll update the PR tomorrow to adjust for the feedback and add tests for this behavior.

> Maybe a good way of doing that is using the `install_rpath` argument to [executable](https://mesonbuild.com/Reference-manual_functions.html#executable)

I think we should test both that, and setting the RPATH by passing linker arguments to Meson.

Reply:
> I slightly disagree. IMO we should be able to change RPATHs set by Meson directly if it doesn't match the behavior we want,

In general such things are Meson-internal implementation details that we don't want to have knowledge of within `meson-python`, so in practice it comes down to the same thing I'd say.

> I don't think this is an unreasonable expectation, but it definitely depends on the project. We should have more documentation regarding this to help users.

I can have a go at this. I'd like to keep it short and lean heavily on what I already wrote for pypackaging-native for concept, and on other tutorials and docs like those of `cibuildwheel` for how to actually go about releasing wheels.